### DISCUSSION TITLE
Autousing fixtures while testing fns which have explicit type-checks?

### DISCUSSION THREAD

**Discussion by xor-xor** [pytest-dev/pytest#9706] [Q&A]
Let's say that I need to test some function, which explicitly checks the type of its input, and I have no control over this behavior. Also, let's say, that I need a fixture for such test:

```python
import pytest


@pytest.fixture()
def my_fixture():
    return 1


def alien_fun(x):
    assert type(x) == int, "sorry, integers only!"
    return x + 1


def test_alien_fun(my_fixture):
    print(type(my_fixture))

    result = alien_fun(my_fixture)

    assert result == 2
```
So far, so good - everything works as expected and I get `<class 'int'>` in the output.

BUT!

If I change this fixture to `autouse=True` (because e.g. I have a lots of tests like the one above and I just don't want to repeat the same declaration over and over):
```python
import pytest


@pytest.fixture(autouse=True)
def my_fixture():
    return 1


def alien_fun(x):
    assert type(x) == int, "sorry, integers only!"
    return x + 1


def test_alien_fun():
    print(type(my_fixture))

    result = alien_fun(my_fixture)

    assert result == 2
```
...then everything breaks with `AssertionError: sorry, integers only!` (well, obviously), because `my_fixture` is now `<class 'function'>`.

Assuming that this is a feature, not a bug, my question is: **what is the "canonical" way of handling such situations?**

EDIT: actually, I've just realized that this type assertion here is irrelevant - this `autouse=True` case will just "explode" on `return x + 1` for exactly the same reason.

**Comment by portyu9** [comment]
Hi @xor-xor 

`autouse=True` makes pytest **execute** the fixture automatically. It does not replace the fixture function's Python name with its return value.

So if you write:

```python
@pytest.fixture(autouse=True)
def my_fixture():
    return 123
```

then inside ordinary module code, `my_fixture` still refers to the fixture function object.

To receive `123`, request the fixture normally:

```python
def test_something(my_fixture):
    assert isinstance(my_fixture, int)
```

Using `autouse=True` is mainly useful when a fixture's **side effects** should happen for every test without adding the fixture argument explicitly.

If you actually need the fixture value, keeping it as an explicit test/fixture argument is the idiomatic approach and also works best with type annotations and IDEs.

Please mark this answer as accepted if it helped, thank you.
